### PR TITLE
Add LRU cache (50 files) to `LinkResolver`

### LINKED ISSUE #992: Add LRU cache to `LinkResolver` to speed up dynamic data source loading
Subgraphs like Melon or Uniswap have a large number of dynamic data sources that were created from a small set of templates. When these subgraphs start and their dynamic data sources are loaded, the same files are resolved via IPFS over and over again.

This process right now can take 10 minutes or more (Melon).

### PR DESCRIPTION
**Pull request by Jannis**
Resolves #992.

This avoids unnecessary roundtrips to IPFS for files that are requested again and again in quick succession, as is the case with data sources created from the same templates.

To not add a dependency on `lru_time_cache` to the `graph` crate, this PR moves the `LinkResolver` implementation to `graph-core`. This requires wrapping `IpfsClient` instead of implementing `LinkResolver` for it directly.

With this, starting a subgraph first logs cache misses:
```
component: SubgraphAssignmentProvider
 subgraph_id: QmRBpsAmUqazecaZ2XCmiLjrEgXJR3JoGyw2kNky6GdZCF
  Jun 07 16:34:22.057 INFO Resolve schema, link: /ipfs/QmebUhZs6i7kCPcLqyP2m6ZEbW69ouTkPQetWdYx4Am9oJ
  Jun 07 16:34:22.057 TRCE IPFS cache miss, hash: QmebUhZs6i7kCPcLqyP2m6ZEbW69ouTkPQetWdYx4Am9oJ
  Jun 07 16:34:22.058 INFO Resolve data source, name: RegistryDataSource
  Jun 07 16:34:22.058 INFO Resolve mapping, link: /ipfs/QmRpvhz37B2rzYQ42GF2LARaCvZ17CFGs7x2z1cGcQ5edp
  Jun 07 16:34:22.058 INFO Resolve ABI, link: /ipfs/QmU2NJjiHMBNxzzuSrFhis24qjR5eCWKZGQGKnGRNptH3u, name: RegistryContract
  Jun 07 16:34:22.059 TRCE IPFS cache miss, hash: QmU2NJjiHMBNxzzuSrFhis24qjR5eCWKZGQGKnGRNptH3u
  Jun 07 16:34:22.059 TRCE IPFS cache miss, hash: QmRpvhz37B2rzYQ42GF2LARaCvZ17CFGs7x2z1cGcQ5edp
  Jun 07 16:34:22.059 INFO Resolve data source, name: EngineDataSource
  Jun 07 16:34:22.060 INFO Resolve mapping, link: /ipfs/QmSnaZn9vNMYWtnkDUeBxvyEyFkydoSdKQQDGpJi7nayCt
  Jun 07 16:34:22.060 INFO Resolve ABI, link: /ipfs/QmTBn21gAsbzL3yfBEaJQKbjCY7Gd6ZjjXt8muJXob61Ba, name: EngineContract
  Jun 07 16:34:22.060 TRCE IPFS cache miss, hash: QmTBn21gAsbzL3yfBEaJQKbjCY7Gd6ZjjXt8muJXob61Ba
  Jun 07 16:34:22.060 TRCE IPFS cache miss, hash: QmSnaZn9vNMYWtnkDUeBxvyEyFkydoSdKQQDGpJi7nayCt
  Jun 07 16:34:22.060 INFO Resolve data source, name: PriceSourceDataSource
  Jun 07 16:34:22.061 INFO Resolve mapping, link: /ipfs/QmVtzBJBNu22bZvqDCMAPihv27iZEuX2KZCrxkxWA7fqWb
  Jun 07 16:34:22.061 INFO Resolve ABI, link: /ipfs/QmQiETkGht3YFtw1vfuT31F993EsKrZddRPEGoB7zoBNnG, name: PriceSourceContract
  Jun 07 16:34:22.061 TRCE IPFS cache miss, hash: QmQiETkGht3YFtw1vfuT31F993EsKrZddRPEGoB7zoBNnG
```

but when its dynamic data sources are loaded, they all hit the cache:

```
omponent: SubgraphAssignmentProvider
 subgraph_id: QmRBpsAmUqazecaZ2XCmiLjrEgXJR3JoGyw2kNky6GdZCF
  Jun 07 16:34:23.048 INFO Resolve data source, name: PolicyManagerDataSource
  Jun 07 16:34:23.048 INFO Resolve mapping, link: /ipfs/QmYpbn1uMvP6XQEuKJtBty7QKgYY8aT8B59WhbFbPQYXW1
  Jun 07 16:34:23.048 INFO Resolve ABI, link: /ipfs/QmUeHnKj8DQqWdgvvNh5hTi4hgsoCQaJSBy9C3rWed3AkM, name: PolicyManagerContract
  Jun 07 16:34:23.049 TRCE IPFS cache hit, hash: QmUeHnKj8DQqWdgvvNh5hTi4hgsoCQaJSBy9C3rWed3AkM
  Jun 07 16:34:23.049 INFO Resolve ABI, link: /ipfs/QmRQMAWbzTrQjhjP1KP7fLKd95hpDTA6yUyYTHHw7D5pAr, name: PolicyContract
  Jun 07 16:34:23.049 TRCE IPFS cache hit, hash: QmRQMAWbzTrQjhjP1KP7fLKd95hpDTA6yUyYTHHw7D5pAr
  Jun 07 16:34:23.049 TRCE IPFS cache hit, hash: QmYpbn1uMvP6XQEuKJtBty7QKgYY8aT8B59WhbFbPQYXW1
  Jun 07 16:34:23.054 INFO Resolve data source, name: FeeManagerDataSource
  Jun 07 16:34:23.055 INFO Resolve mapping, link: /ipfs/QmRk7Ch7cPLbsGF6vY91ywWtF4BNuMQM7b5b1KLhgb22mt
  Jun 07 16:34:23.055 INFO Resolve ABI, link: /ipfs/QmTgmfrWkBzL6fbTMnRvht79Lv4LYWZU3RkhdPbZ1Azwrf, name: FeeManagerContract
  Jun 07 16:34:23.055 TRCE IPFS cache hit, hash: QmTgmfrWkBzL6fbTMnRvht79Lv4LYWZU3RkhdPbZ1Azwrf
  Jun 07 16:34:23.055 INFO Resolve ABI, link: /ipfs/QmZtu75zpxjhFQUZdAHQ9aWBNEpQ3C5GXymTid2Kabt5GP, name: ManagementFeeContract
  Jun 07 16:34:23.055 TRCE IPFS cache hit, hash: QmZtu75zpxjhFQUZdAHQ9aWBNEpQ3C5GXymTid2Kabt5GP
  Jun 07 16:34:23.056 INFO Resolve ABI, link: /ipfs/QmYRewDG2p5XM1EC6nan1LcutMYYFK31bGba2YBiDXMtPy, name: PerformanceFeeContract
  Jun 07 16:34:23.056 TRCE IPFS cache hit, hash: QmYRewDG2p5XM1EC6nan1LcutMYYFK31bGba2YBiDXMtPy
  Jun 07 16:34:23.056 TRCE IPFS cache hit, hash: QmRk7Ch7cPLbsGF6vY91ywWtF4BNuMQM7b5b1KLhgb22mt
  Jun 07 16:34:23.065 INFO Resolve data source, name: AccountingDataSource
  Jun 07 16:34:23.065 INFO Resolve mapping, link: /ipfs/QmYCPvPangFPhU5qG784zur2Y8gz7cgDStcbVQ4aRddbH6
  Jun 07 16:34:23.065 INFO Resolve ABI, link: /ipfs/QmeWyLGruNRtPVLMX8gEq9ASe1tkQCF6FfvxW45AqLigb5, name: AccountingContract
  Jun 07 16:34:23.066 TRCE IPFS cache hit, hash: QmeWyLGruNRtPVLMX8gEq9ASe1tkQCF6FfvxW45AqLigb5
  Jun 07 16:34:23.066 TRCE IPFS cache hit, hash: QmYCPvPangFPhU5qG784zur2Y8gz7cgDStcbVQ4aRddbH6
  Jun 07 16:34:23.074 INFO Resolve data source, name: FeeManagerDataSource
  Jun 07 16:34:23.074 INFO Resolve mapping, link: /ipfs/QmRk7Ch7cPLbsGF6vY91ywWtF4BNuMQM7b5b1KLhgb22mt
  Jun 07 16:34:23.075 INFO Resolve ABI, link: /ipfs/QmTgmfrWkBzL6fbTMnRvht79Lv4LYWZU3RkhdPbZ1Azwrf, name: FeeManagerContract
  Jun 07 16:34:23.075 TRCE IPFS cache hit, hash: QmTgmfrWkBzL6fbTMnRvht79Lv4LYWZU3RkhdPbZ1Azwrf
  Jun 07 16:34:23.075 INFO Resolve ABI, link: /ipfs/QmZtu75zpxjhFQUZdAHQ9aWBNEpQ3C5GXymTid2Kabt5GP, name: ManagementFeeContract
  Jun 07 16:34:23.075 TRCE IPFS cache hit, hash: QmZtu75zpxjhFQUZdAHQ9aWBNEpQ3C5GXymTid2Kabt5GP
  Jun 07 16:34:23.075 INFO Resolve ABI, link: /ipfs/QmYRewDG2p5XM1EC6nan1LcutMYYFK31bGba2YBiDXMtPy, name: PerformanceFeeContract
  Jun 07 16:34:23.076 TRCE IPFS cache hit, hash: QmYRewDG2p5XM1EC6nan1LcutMYYFK31bGba2YBiDXMtPy
  Jun 07 16:34:23.076 TRCE IPFS cache hit, hash: QmRk7Ch7cPLbsGF6vY91ywWtF4BNuMQM7b5b1KLhgb22mt
  Jun 07 16:34:23.083 INFO Resolve data source, name: SharesDataSource
  Jun 07 16:34:23.083 INFO Resolve mapping, link: /ipfs/QmVpECxecVWXg8dbZeBztX9ECsyhsEM4u3dVEDGHsn5T9D
  Jun 07 16:34:23.083 INFO Resolve ABI, link: /ipfs/QmWmHBCbXagiVxSUXSC7YX3zvf2kU8ppTRA5bd9uQ9hMYG, name: SharesContract
  Jun 07 16:34:23.084 TRCE IPFS cache hit, hash: QmWmHBCbXagiVxSUXSC7YX3zvf2kU8ppTRA5bd9uQ9hMYG
  Jun 07 16:34:23.084 TRCE IPFS cache hit, hash: QmVpECxecVWXg8dbZeBztX9ECsyhsEM4u3dVEDGHsn5T9D
  Jun 07 16:34:23.089 INFO Resolve data source, name: SharesDataSource
  Jun 07 16:34:23.089 INFO Resolve mapping, link: /ipfs/QmVpECxecVWXg8dbZeBztX9ECsyhsEM4u3dVEDGHsn5T9D
  Jun 07 16:34:23.089 INFO Resolve ABI, link: /ipfs/QmWmHBCbXagiVxSUXSC7YX3zvf2kU8ppTRA5bd9uQ9hMYG, name: SharesContract
  Jun 07 16:34:23.090 TRCE IPFS cache hit, hash: QmWmHBCbXagiVxSUXSC7YX3zvf2kU8ppTRA5bd9uQ9hMYG
  Jun 07 16:34:23.090 TRCE IPFS cache hit, hash: QmVpECxecVWXg8dbZeBztX9ECsyhsEM4u3dVEDGHsn5T9D
```

### CODE REVIEW

**Review by leoyvens — CHANGES_REQUESTED** [review]
Besides the minor comment, I'm concerned about the total size of the cache. 50*MAX_IPFS_MAP_FILE_SIZE is 12GB, this could hog memory and even be an attack vector. We have to limit the size further somehow, I propose that we only cache small files, even though we still allow fetching large files.

File: runtime/wasm/src/host_exports.rs [hunk 1]
```diff
@@ -352,11 +352,12 @@ where
 
     pub(crate) fn ipfs_cat(
         &self,
+        module: &WasmiModule<E, L, S, U>,
```
[the remark this answers is not in the document]
Comment (Jannis) [reply runtime/wasm/src/host_exports.rs]: Yep

**Comment by Jannis** [comment]
@leodasvacas That's a good point. I'll update the PR and also make the cache size configurable.

**Comment by Jannis** [comment]
@leodasvacas Updated. See the latest commit (https://github.com/graphprotocol/graph-node/pull/993/commits/7c91bb4895bb6c6dd5d1fd109181716e325c4b41) in this PR:

> **core: Add limits for IPFS cache and cache file sizes**
>
> The maximum cache capacity is now controlled by the `GRAPH_MAX_IPFS_CACHE_SIZE` environment variable. Only files smaller than or equal to `GRAPH_MAX_IPFS_CACHE_FILE_SIZE` are cached. Larger files are always loaded from IPFS.

**Comment by Jannis** [comment]
I still need to add these vars to `docs/`.

**Comment by Jannis** [comment]
...and that is now done as well.

**Comment by leodasvacas** [comment]
Changes look good, but according to CI this is not compiling.

**Comment by Jannis** [comment]
@leodasvacas Updated once more. Should pass the test suite now.

**Review by leoyvens — APPROVED** [review]

**Pull request merged**